lib/jellyfish_web/socket.ex

```diff
@@ -33,46 +37,97 @@ defmodule JellyfishWeb.Socket do
 
       {:ok, state}
     else
-      {:ok, :connected} ->
+      params: :error ->
+        Logger.warn(
+          "No room_id or peer_id in connection params, ignoring incoming WebSocket connection"
+        )
+
+        {:error, :no_params}
+
+      peer: {:ok, _status} ->
         Logger.warn("""
         WebSocket connection for peer #{inspect(state.params["peer_id"])} in room \
         #{inspect(state.params["room_id"])} already exists, rejected
         """)
 
         {:error, :already_connected}
 
-      {:error, :room_not_found} ->
+      room: {:error, :room_not_found} ->
         Logger.warn("""
         Room #{inspect(state.params["room_id"])} not found, \
         ignoring incoming WebSocket connection
         """)
 
         {:error, :room_not_found}
 
-      {:error, :peer_not_found} ->
+      peer: {:error, :peer_not_found} ->
         Logger.warn("""
         Peer #{inspect(state.params["peer_id"])} not found in room \
         #{inspect(state.params["room_id"])}, ignoring incoming WebSocket connection
         """)
 
         {:error, :peer_not_found}
-
-      :error ->
-        Logger.warn(
-          "No room_id/peer_id in connection params, ignoring incoming WebSocket connection"
-        )
-
-        {:error, :no_params}
     end
   end
 
   @impl true
   def init(state) do
-    Room.set_peer_connected(state.room_pid, state.peer_id)
-    {:ok, state}
+    Room.set_peer_authenticating(state.room_pid, state.peer_id)
+
+    if Application.fetch_env!(:jellyfish, :server_auth) do
+      {:ok, Map.put(state, :authenticated, false)}
+    else
+      Room.set_peer_connected(state.room_pid, state.peer_id)
+      {:ok, Map.put(state, :authenticated, true)}
+    end
   end
 
   @impl true
+  def handle_in({encoded_message, [opcode: :text]}, %{authenticated: false} = state) do
+    case Jason.decode(encoded_message) do
+      {:ok, %{"type" => "controlMessage", "data" => %{"type" => "authRequest", "token" => token}}} ->
+        with {:ok, peer_id} <-
+               Phoenix.Token.verify(
+                 JellyfishWeb.Endpoint,
+                 Application.get_env(:jellyfish, :auth_salt),
+                 token,
+                 max_age: Application.get_env(:jellyfish, :jwt_max_age)
+               ),
+             :ok <- if(peer_id == state.peer_id, do: :ok, else: {:error, :invalid_peer_id}) do
```

Comment:
What do you think about moving this `if` to a function like `verifey_peer_id` or something like that?